flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+				}
+			}
+		});
+	}
+
+	@Override
+	public void onModified(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onDeleted(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onError(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@VisibleForTesting
+	ConcurrentMap<ResourceID, KubernetesWorkerNode> getWorkerNodeMap() {
+		return workerNodeMap;
+	}
+
+	private void getWorkerNodesFromPreviousAttempts() throws Exception {
+		final List<KubernetesPod> podList = kubeClient.getPodsWithLabels(getTaskManagerLabels());
+		for (KubernetesPod pod : podList) {
+			final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+			workerNodeMap.put(worker.getResourceID(), worker);
+			final long attempt = worker.getAttempt();
+			if (attempt > currentMaxAttemptId.get()) {
+				currentMaxAttemptId.set(attempt);
+			}
+		}
+
+		log.info("Recovered {} pods from previous attempts, current attempt id is {}.",
+			workerNodeMap.size(),
+			currentMaxAttemptId.addAndGet(1));
+	}
+
+	private void requestKubernetesPod() {
+		Preconditions.checkNotNull(this.kubeClient);
+
+		numPendingPodRequests++;
+
+		log.info("Requesting new TaskExecutor pod with <{},{}>. Number pending requests {}.",
+			defaultTaskManagerMemoryMB,
+			defaultCpus,
+			numPendingPodRequests);
+
+		final String podName = String.format(
+			TASK_MANAGER_POD_FORMAT,
+			clusterId,
+			currentMaxAttemptId.get(),
+			currentMaxPodId.addAndGet(1));
+
+		try {
+			final HashMap<String, String> env = new HashMap<>();
+			env.put(Constants.ENV_FLINK_POD_NAME, podName);
+
+			final TaskManagerPodParameter parameter = new TaskManagerPodParameter(
+				podName,
+				getTaskManagerStartCommand(podName),
+				defaultTaskManagerMemoryMB,
+				defaultCpus,
+				env);
+
+			this.kubeClient.createTaskManagerPod(parameter);
+		} catch (Exception e) {
+			this.kubeClient.handleException(e);
+			throw new FlinkRuntimeException("Could not start new worker");
+		}
+	}
+
+	/**
+	 * Request new pod if pending pods cannot satisfies pending slot requests.
+	 */
+	private void requestKubernetesPodIfRequired() {
+		final int requiredTaskManagerSlots = getNumberRequiredTaskManagerSlots();
+		final int pendingTaskManagerSlots = numPendingPodRequests * numberOfTaskSlots;
+
+		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+			requestKubernetesPod();
+		}
+	}
+
+	private void removePodIfTerminated(KubernetesPod pod) {
+		if (pod.isTerminated()) {
+			kubeClient.stopPod(pod.getName());
+			final KubernetesWorkerNode kubernetesWorkerNode = workerNodeMap.remove(new ResourceID(pod.getName()));
+			if (kubernetesWorkerNode != null) {
+				requestKubernetesPodIfRequired();
+			}
+		}
+	}
+
+	private List<String> getTaskManagerStartCommand(String podName) {
```

Comment:
This command list can be constructed only once, maybe during constructor or `initialize()`, then we can avoid executing this repeatedly while requesting every pod.

Reply:
Make sense. I will add `taskManagerStartCommand` and initialize it in the constructor.